### DISCUSSION TITLE
Automatic Calibration of Anchor Location for LPS [Proposal]

### DISCUSSION THREAD

**Discussion by magicbycalvin** [bitcraze/discussions#62] [Ideas]
As suggested [here](https://github.com/orgs/bitcraze/discussions/51#discussioncomment-2839443), I would like to propose the implementation of automatic location calibration of the LPS anchors. For the present system, it is necessary to manually measure and record the 3D location of each anchor. This can introduce measurement uncertainties (which may not be too big of an issue since it has been mentioned that a deviation of up to +/- 5 cm shouldn't cause much of a problem) and is quite an inconvenience. Furthermore, it is infeasible to quickly move the LPS to a different location without a team of people familiar with the system. Fortunately, there is plenty of work that has been done with regards to state estimation and while implementing automatic calibration will by no means be trivial, it should be possible.

@knmcguire has mentioned that they will share some relevant publications regarding this idea. As this system is tangentially related to my research, I will be happy to spend some free time looking into the solution as well. Any helpful contributions from others will be very much appreciated!

Thank you for your time!

**Comment by krichardsson** [comment]
I did some experiments some years ago on this topic. I think it can be done but I did not manage to reach a point where I thought the functionality justified the extra complexity in the code base. I only spent a day or so on the problem though :-)

A starting point could be the work by Mike Hamer (http://mikehamer.info/uwb-system/) back in 2016. Mike also wrote the initial version of the kalman filter that is used in the Crazyflie.

**Reply by knmcguire** [reply]
Ah yes, this was exactly the work that I wanted to share with @magicbycalvin. thanks kristoffer! Hope this will be a good starting point

**Reply by magicbycalvin** [reply]
Excellent, thank you for sharing! I will take a look at this during lunch today. I was thinking that we might be able to leverage the excellent GTSAM library which smooths, rather than filters the sensor data which may result in better localization. They also already have pure range measurement factor classes which should be exactly what we're looking for.

**Comment by PioBeat** [comment]
I recently shared a tool called XR-PALS in a separate discussion here, but thought it could be relevant to this topic too:
https://github.com/orgs/bitcraze/discussions/1995

It lets you visually place and measure anchor positions in 3D via a Meta Quest 3.
I think this might be somewhat related to your idea — you could consider it a semi-automatic location calibration approach for LPS anchors using mixed reality.

Personally, it simplified my setup in different rooms. What do you think?
